File: GSEA.and.intersections/run_enrichr_DEG_log2FC0.2.py
#!/usr/bin/env python3
"""
Enrichr (gseapy) + barplots in ONE script.
  - DEG mode (default): tables/DEG_log2FC0.2/genes_only/*.txt
      -> tables/DEG_log2FC0.2/enrichr_results/*_enrichr.csv + barplots/
  - genes_in_common mode: genes_in_common/*_genes_in_common.txt
      -> genes_in_common/enrichr_results/*_enrichr.csv + barplots/

Usage:
  python3 run_enrichr_DEG_log2FC0.2.py              # DEG log2FC0.2
  python3 run_enrichr_DEG_log2FC0.2.py --genes-in-common
"""

import sys
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt

ENRICHR_LIBRARIES = [
    "KEGG_2021_Human",
    "Reactome_2022",
    "MSigDB_Hallmark_2020",
    "WikiPathway_2023_Human",
    "GO_Biological_Process_2023",
]

ADJ_PVAL = 0.2
TOP_N = 25
LIBRARIES = [
    ("WikiPathway_2023_Human", "WikiPathways"),
    ("MSigDB_Hallmark_2020", "Hallmark"),
    ("GO_Biological_Process_2023", "GO_BP"),
    ("Reactome_2022", "Reactome"),
    ("KEGG_2021_Human", "KEGG"),
]


def plot_enrichr_csv(csv_path: Path, barplots_dir: Path) -> None:
    """Barplots for one combined Enrichr CSV (Gene_set + Term + Adjusted P-value)."""
    barplots_dir.mkdir(parents=True, exist_ok=True)
    df = pd.read_csv(csv_path)
    pcol = "Adjusted P-value" if "Adjusted P-value" in df.columns else "P-value"
    if pcol not in df.columns or "Gene_set" not in df.columns or "Term" not in df.columns:
        print("  Skip plots:", csv_path.name, "(missing columns)")
        return
    base_name = csv_path.stem.replace("_enrichr", "")
    for gene_set_value, short_name in LIBRARIES:
        sub = df[(df["Gene_set"] == gene_set_value) & (df[pcol] < ADJ_PVAL)].copy()
        if sub.empty:
            continue
        sub = sub.sort_values(pcol).head(TOP_N)
        pvals = np.clip(sub[pcol].values, 1e-300, 1.0)
        y = -np.log10(pvals)
        terms = sub["Term"].tolist()
        terms_short = [t[:60] + "..." if len(t) > 60 else t for t in terms]
        n = len(terms_short)
        fig, ax = plt.subplots(figsize=(8, max(5, n * 0.35)))
        y_pos = np.arange(n)[::-1]
        ax.barh(y_pos, y, color="steelblue", edgecolor="navy", linewidth=0.5)
        ax.set_yticks(y_pos)
        ax.set_yticklabels(terms_short, fontsize=8)
        ax.set_xlabel("-log10(Adjusted P-value)")
        ax.set_title(f"{base_name} — {short_name}\n(Adj P < {ADJ_PVAL}, top {n} terms)")
        ax.axvline(-np.log10(ADJ_PVAL), color="gray", linestyle="--", linewidth=1, alpha=0.8)
        ax.spines["top"].set_visible(False)
        ax.spines["right"].set_visible(False)
        plt.tight_layout()
        plt.savefig(barplots_dir / f"{base_name}_{short_name}_barplot.png", dpi=150, bbox_inches="tight")
        plt.close()


def run_enrichr_and_plot(txt_path: Path, out_dir: Path) -> None:
    try:
        import gseapy as gp
    except ImportError:
        print("ERROR: pip install gseapy")
        sys.exit(1)
    genes = [g.strip() for g in txt_path.read_text().strip().splitlines() if g.strip()]
    if not genes:
        print(txt_path.name, ": empty, skip")
        return
    print(txt_path.name, f"({len(genes)} genes)...", end=" ", flush=True)
    all_dfs = []
    for lib in ENRICHR_LIBRARIES:
        enr = gp.enrichr(genes, gene_sets=lib, organism="Human", outdir=None, no_plot=True, verbose=False)
        res = enr.results.copy()
        res["Gene_set"] = lib
        all_dfs.append(res)
    combined = pd.concat(all_dfs, ignore_index=True)
    out_csv = out_dir / f"{txt_path.stem}_enrichr.csv"
    combined.to_csv(out_csv, index=False)
    print("->", out_csv.name)
    # Barplots next to tables
    barplots_dir = out_dir / "barplots"
    plot_enrichr_csv(out_csv, barplots_dir)


def main():
    genes_in_common = "--genes-in-common" in sys.argv
    base = Path(__file__).resolve().parent.parent

    if genes_in_common:
        genes_dir = base / "genes_in_common"
        out_dir = genes_dir / "enrichr_results"
        pattern = "*_genes_in_common.txt"
    else:
        genes_dir = base / "tables" / "DEG_log2FC0.2" / "genes_only"
        out_dir = base / "tables" / "DEG_log2FC0.2" / "enrichr_results"

    if not genes_dir.is_dir():
        print("ERROR: Not found:", genes_dir)
        sys.exit(1)
    out_dir.mkdir(parents=True, exist_ok=True)

    if genes_in_common:
        txt_files = sorted(genes_dir.glob(pattern))
    else:
        txt_files = sorted(genes_dir.glob("*.txt"))

    if not txt_files:
        print("No input files in", genes_dir)
        sys.exit(1)

    print("Mode:", "genes_in_common" if genes_in_common else "DEG_log2FC0.2")
    print("Input:", genes_dir)
    print("Output:", out_dir)
    print("Barplots:", out_dir / "barplots", "\n")

    for path in txt_files:
        run_enrichr_and_plot(path, out_dir)

    print("\nDone. Tables + barplots in:", out_dir)


if __name__ == "__main__":
    main()

File: GSEA.and.intersections/run_enrichR_DEG_log2FC0.2.R
# =============================================================================
# Enrichr + barplots in ONE script (merged run_enrichR_DEG_log2FC0.2.R + run_enrichR_genes_in_common.R)
# =============================================================================
# Set MODE to:
#   "DEG"            -> tables/DEG_log2FC0.2/genes_only/*.txt
#   "genes_in_common" -> genes_in_common/*_genes_in_common.txt (also terms + Hallmark/KEGG folders)
# Run from project root:  Rscript scripts_FGSEA/run_enrichR_all_in_one.R
# =============================================================================

MODE <- "DEG"   # "DEG" or "genes_in_common"

library(enrichR)
library(readr)
library(dplyr)
library(ggplot2)

base_dir <- normalizePath(".", winslash = "/")

if (MODE == "DEG") {
  genes_dir <- file.path(base_dir, "tables", "DEG_log2FC0.2", "genes_only")
  out_dir <- file.path(base_dir, "tables", "DEG_log2FC0.2", "enrichr_results_R")
  input_pattern <- "\\.txt$"
  use_hall_kegg_dirs <- FALSE
} else {
  genes_dir <- file.path(base_dir, "genes_in_common")
  out_dir <- file.path(base_dir, "genes_in_common", "enrichr_results_R")
  input_pattern <- "_genes_in_common\\.txt$"
  use_hall_kegg_dirs <- TRUE
}

tables_dir <- file.path(out_dir, "tables")
plots_dir <- file.path(out_dir, "plots")
terms_dir <- file.path(out_dir, "terms_AdjP_below_0.2")
dir.create(out_dir, showWarnings = FALSE)
dir.create(tables_dir, showWarnings = FALSE)
dir.create(plots_dir, showWarnings = FALSE)
if (use_hall_kegg_dirs) {
  dir.create(terms_dir, showWarnings = FALSE)
  hall_dir <- file.path(genes_dir, "enrichr_Hallmark_only")
  kegg_dir <- file.path(genes_dir, "enrichr_KEGG_only")
  dir.create(hall_dir, showWarnings = FALSE)
  dir.create(kegg_dir, showWarnings = FALSE)
}

enrichr_databases <- c(
  "KEGG_2021_Human",
  "Reactome_2022",
  "MSigDB_Hallmark_2020",
  "WikiPathway_2023_Human",
  "GO_Biological_Process_2023"
)
adj_p_cutoff <- 0.2
top_n_plot <- 25

txt_files <- list.files(genes_dir, pattern = input_pattern, full.names = TRUE)
txt_files <- sort(txt_files)
if (length(txt_files) == 0) stop("No input files in ", genes_dir)

cat("MODE:", MODE, "\n")
cat("Input:", genes_dir, "\n")
cat("Output:", out_dir, "\n")
cat("Databases:", paste(enrichr_databases, collapse = ", "), "\n\n")

for (f in txt_files) {
  if (MODE == "genes_in_common") {
    base_name <- gsub("_genes_in_common\\.txt$", "", basename(f))
  } else {
    base_name <- sub("\\.txt$", "", basename(f))
  }
  cat("Processing:", base_name, "\n")
  gene_list <- unique(trimws(readLines(f)))
  gene_list <- gene_list[nzchar(gene_list)]
  if (length(gene_list) == 0) { cat("  Skip: no genes.\n"); next }
  cat("  Genes:", length(gene_list), "\n")

  enrichr_results <- tryCatch(enrichr(gene_list, enrichr_databases),
    error = function(e) { cat("  Error:", conditionMessage(e), "\n"); NULL })
  if (is.null(enrichr_results)) next

  if (use_hall_kegg_dirs && nrow(bind_rows(lapply(names(enrichr_results), function(db) {
    df <- enrichr_results[[db]]; if (nrow(df) == 0) return(NULL); df$Gene_set <- db; df
  }), .id = NULL)) > 0) {
    all_res <- bind_rows(lapply(names(enrichr_results), function(db) {
      df <- enrichr_results[[db]]
      if (nrow(df) == 0) return(NULL)
      df$Gene_set <- db
      df
    }))
    if ("Adjusted P-value" %in% names(all_res) && !"Adjusted.P.value" %in% names(all_res))
      all_res$Adjusted.P.value <- all_res[["Adjusted P-value"]]
    below <- all_res %>% filter(Adjusted.P.value < adj_p_cutoff) %>% arrange(Gene_set, Adjusted.P.value)
    if (nrow(below) > 0) {
      write_csv(below, file.path(terms_dir, paste0(base_name, "_terms_AdjP_below_0.2.csv")))
      below_hall <- below %>% filter(grepl("Hallmark", Gene_set, ignore.case = TRUE))
      if (nrow(below_hall) > 0)
        write_csv(below_hall, file.path(hall_dir, paste0(base_name, "_terms_AdjP_below_0.2_Hallmark_only.csv")))
      below_kegg <- below %>% filter(grepl("KEGG", Gene_set, ignore.case = TRUE))
      if (nrow(below_kegg) > 0)
        write_csv(below_kegg, file.path(kegg_dir, paste0(base_name, "_terms_AdjP_below_0.2_KEGG_only.csv")))
    }
  }

  for (db in names(enrichr_results)) {
    result_df <- enrichr_results[[db]]
    if (nrow(result_df) == 0) next
    if (!"Adjusted.P.value" %in% names(result_df) && "Adjusted P-value" %in% names(result_df))
      result_df$Adjusted.P.value <- result_df[["Adjusted P-value"]]
    significant <- result_df %>%
      filter(Adjusted.P.value < adj_p_cutoff) %>%
      arrange(Adjusted.P.value) %>%
      head(top_n_plot)
    if (nrow(significant) == 0) next

    db_short <- switch(db,
      "WikiPathway_2023_Human" = "WikiPathways",
      "MSigDB_Hallmark_2020" = if (use_hall_kegg_dirs) "MSigDB_Hallmark" else "Hallmark",
      "GO_Biological_Process_2023" = if (use_hall_kegg_dirs) "GO_Biological_Process" else "GO_BP",
      "Reactome_2022" = "Reactome",
      "KEGG_2021_Human" = "KEGG", db)

    write_csv(significant, file.path(tables_dir, paste0(base_name, "_", db_short, ".csv")))
    if (use_hall_kegg_dirs && db_short == "MSigDB_Hallmark")
      write_csv(significant, file.path(hall_dir, paste0(base_name, "_", db_short, ".csv")))
    if (use_hall_kegg_dirs && db_short == "KEGG")
      write_csv(significant, file.path(kegg_dir, paste0(base_name, "_", db_short, ".csv")))

    plot_data <- significant %>%
      mutate(
        Term_short = ifelse(nchar(Term) > 60, paste0(substr(Term, 1, 57), "..."), Term),
        log_p = -log10(pmax(Adjusted.P.value, 1e-300))
      )
    p <- ggplot(plot_data, aes(x = log_p, y = reorder(Term_short, log_p))) +
      geom_col(fill = "steelblue", alpha = 0.8) +
      geom_vline(xintercept = -log10(adj_p_cutoff), linetype = "dashed", color = "gray40") +
      labs(title = paste0(base_name, " — ", db_short),
           subtitle = paste0("Adj P < ", adj_p_cutoff, ", top ", nrow(plot_data), " terms"),
           x = "-log10(Adjusted P-value)", y = "") +
      theme_minimal() +
      theme(axis.text.y = element_text(size = 8),
            plot.background = element_rect(fill = "white", color = NA),
            panel.background = element_rect(fill = "white", color = NA))
    plot_path <- file.path(plots_dir, paste0(base_name, "_", db_short, "_barplot.png"))
    ggsave(plot_path, p, width = 10, height = max(5, nrow(plot_data) * 0.3), dpi = 150, bg = "white")
    if (use_hall_kegg_dirs && db_short == "MSigDB_Hallmark")
      file.copy(plot_path, file.path(hall_dir, paste0(base_name, "_", db_short, "_barplot.png")), overwrite = TRUE)
    if (use_hall_kegg_dirs && db_short == "KEGG")
      file.copy(plot_path, file.path(kegg_dir, paste0(base_name, "_", db_short, "_barplot.png")), overwrite = TRUE)
    cat("  ", db_short, ":", nrow(significant), "terms -> table + barplot\n", sep = "")
  }
  cat("\n")
}

cat("Done. Results in:", out_dir, "\n")
cat("  tables/  plots/\n")
if (use_hall_kegg_dirs) cat("  terms_AdjP_below_0.2/  enrichr_Hallmark_only/  enrichr_KEGG_only/\n")
